io/io/src/TFileMerger.cxx

```diff
@@ -608,89 +608,31 @@ Bool_t TFileMerger::MergeRecursive(TDirectory *target, TList *sourcelist, Int_t
-                           obj->GetName(), key->GetName());
-                  }
+               if (cl->GetMethodWithPrototype("Merge", "TCollection*,TFileMergeInfo*")) {
+                  listHargs.Form("(TCollection*)0x%lx,(TFileMergeInfo*)0x%lx",
+                                  (ULong_t)&listH, (ULong_t)&info);
+               } else if (cl->GetMethodWithPrototype("Merge", "TCollection*")) {
+                  listHargs.Form("((TCollection*)0x%lx)", (ULong_t)&listH);
                } else {
-                  while (nextsource) {
-                     // make sure we are at the correct directory level by cd'ing to path
-                     TDirectory *ndir = nextsource->GetDirectory(path);
-                     if (ndir) {
-                        // For consistency (and persformance), we reset the MustCleanup be also for those
-                        // 'key' retrieved indirectly.
-                        //ndir->ResetBit(kMustCleanup);
-                        ndir->cd();
-                        TKey *key2 = (TKey*)ndir->GetListOfKeys()->FindObject(key->GetName());
-                        if (key2) {
-                           TObject *hobj = key2->ReadObj();
-                           if (!hobj) {
-                              Info("MergeRecursive", "could not read object for key {%s, %s}; skipping file %s",
-                                   key->GetName(), key->GetTitle(), nextsource->GetName());
-                              nextsource = (TFile*)sourcelist->After(nextsource);
-                              continue;
-                           }
-                           // Set ownership for collections
-                           if (hobj->InheritsFrom(TCollection::Class())) {
-                              ((TCollection*)hobj)->SetOwner();
-                           }
-                           hobj->ResetBit(kMustCleanup);
-                           listH.Add(hobj);
-                           Int_t error = 0;
-                           obj->Execute("Merge", listHargs.Data(), &error);
-                           info.fIsFirst = kFALSE;
-                           if (error) {
-                              Error("MergeRecursive", "calling Merge() on '%s' with the corresponding object in '%s'",
-                                    obj->GetName(), nextsource->GetName());
-                           }
-                           listH.Delete();
-                        }
-                     }
-                     nextsource = (TFile*)sourcelist->After( nextsource );
-                  }
-                  // Merge the list, if still to be done
-                  if (info.fIsFirst) {
-                     Int_t error = 0;
-                     obj->Execute("Merge", listHargs.Data(), &error);
-                     info.fIsFirst = kFALSE;
-                     listH.Delete();
-                  }
+                  // skip any unhandled TObjects
```

Comment:
This part seems to be new (backward incompatible) change, is that really intended?

Reply:
I thought this is in effect what is already happening, since the only remaining case after this one is the "can't be merged case" 

https://github.com/root-project/root/blob/03bab5b4834ee3b5bda5599581c20bff063f0ef5/io/io/src/TFileMerger.cxx#L681-L684

Reply:
Is it that the new log output is backwards incompatible?

Reply:
It is the behavior.  After setting canBeMerged the objects still ends up executing one of branches in lines 724 to 744.
Effectively the non-mergeable objects are all written as is (i.e. multiple objects instead of a single merged one).

Reply:
I see, thank you for pointing that out. That was certainly not intended!